### ISSUE TITLE
Internal Error occured during "Ready state poller"

### ISSUE THREAD

**Issue by Flaw101** [spring-projects/spring-ide#182]
What steps will reproduce the problem?
1. Launching Spring Boot App from the Boot Dashboard. I haven't found any app that doesn't throw this exception.

-- Error Details --
```
Date: Wed Aug 23 13:50:00 BST 2017
Message: An internal error occurred during: "Ready state poller".
Severity: Error
Product: Spring Tool Suite 3.9.0.201707061823-RELEASE (org.springsource.sts.ide)
Plugin: org.eclipse.core.jobs
Session Data:
eclipse.buildId=3.9.0.201707061823-RELEASE-e47
java.version=1.8.0_131
java.vendor=Oracle Corporation
BootLoader constants: OS=macosx, ARCH=x86_64, WS=cocoa, NL=en_US
Framework arguments:  -product org.springsource.sts.ide 
Command-line arguments:  -os macosx -ws cocoa -arch x86_64 -product org.springsource.sts.ide 

```
Exception Stack Trace:
```
com.google.common.util.concurrent.UncheckedExecutionException: java.lang.NullPointerException
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2199)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3932)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3936)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4806)
	at org.springframework.ide.eclipse.boot.dash.model.LocalCloudServiceDashElement.getLaunchConfigs(LocalCloudServiceDashElement.java:144)
	at org.springframework.ide.eclipse.boot.dash.model.AbstractLaunchConfigurationsDashElement$4.stateChanged(AbstractLaunchConfigurationsDashElement.java:434)
	at org.springframework.ide.eclipse.boot.dash.model.AbstractLaunchConfigurationsDashElement$4.stateChanged(AbstractLaunchConfigurationsDashElement.java:1)
	at org.springframework.ide.eclipse.boot.dash.util.RunStateTracker.updateOwnerStatesAndFireEvents(RunStateTracker.java:216)
	at org.springframework.ide.eclipse.boot.dash.util.RunStateTracker.access$0(RunStateTracker.java:208)
	at org.springframework.ide.eclipse.boot.dash.util.RunStateTracker$1.gotValue(RunStateTracker.java:151)
	at org.springframework.ide.eclipse.boot.dash.util.RunStateTracker$1.gotValue(RunStateTracker.java:1)
	at org.springsource.ide.eclipse.commons.livexp.core.LiveExpression.changed(LiveExpression.java:114)
	at org.springsource.ide.eclipse.commons.livexp.core.LiveExpression.refresh(LiveExpression.java:71)
	at org.springsource.ide.eclipse.commons.livexp.core.LiveVariable.setValue(LiveVariable.java:46)
	at org.springframework.ide.eclipse.boot.dash.util.AbstractPollingAppReadyStateMonitor$1.run(AbstractPollingAppReadyStateMonitor.java:51)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:56)
Caused by: java.lang.NullPointerException
	at org.eclipse.debug.internal.core.LaunchManager.isValidLaunchConfigurationName(LaunchManager.java:2712)
	at org.eclipse.debug.internal.core.LaunchConfigurationType.newInstance(LaunchConfigurationType.java:473)
	at org.springframework.ide.eclipse.boot.launch.cli.CloudCliServiceLaunchConfigurationDelegate.createLaunchConfig(CloudCliServiceLaunchConfigurationDelegate.java:156)
	at org.springframework.ide.eclipse.boot.dash.model.LocalCloudServiceDashElement$1.load(LocalCloudServiceDashElement.java:68)
	at org.springframework.ide.eclipse.boot.dash.model.LocalCloudServiceDashElement$1.load(LocalCloudServiceDashElement.java:1)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3522)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2315)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2278)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2193)
	... 15 more
```

**Comment by BoykoAlex** [comment]
Any idea what’s the boot and cloud cli versions you have STS pointed to? You should be able to find that out via Preferences -> Boot -> CLI… I suspect that the result of parsing available services from Cloud CLI command is not good… i.e. `null` is passed as a service id and hence this NPE.
Also I’d be interested in seeing your STS logs… or alternatively if you find the Spring Boot CLI folder STS points too (if it’s not a remote URL) then the output of this command `spring cloud --list`

**Comment by Flaw101** [comment]
The boot CLI was set to the default with STS, `1.5.4.RELEASE` The cloud CLI was an extension (looks?) just be packaged as part of the boot cli? Set to `1.3.3`  I Downloaded the `1.5.6.RELEASE` manually, 
Installing the `1.5.6.RELEASE` with the Cloud CLI `1.3.4.RELEASE` the error initially never happened, but after a restart it appeared with this. These apps are also pretty much empty shells, a few SQS beans one is a carbon copy of the other just pointing to a different queue. One starts and the other does not. 

```
Spring Cloud CLI v1.3.4.RELEASE
15:12:25.804 [main] DEBUG org.springframework.cloud.launcher.cli.LauncherCommand - resolved URIs [file:/Users/n0225684/.m2/repository/org/springframework/cloud/launcher/spring-cloud-launcher-deployer/1.3.4.RELEASE/spring-cloud-launcher-deployer-1.3.4.RELEASE.jar]
15:12:25.883 [main] DEBUG org.springframework.cloud.launcher.cli.LauncherCommand - resolved URIs [file:/Users/n0225684/.m2/repository/org/springframework/cloud/launcher/spring-cloud-launcher-deployer/1.3.4.RELEASE/spring-cloud-launcher-deployer-1.3.4.RELEASE.jar]
15:13:42.197 [main] ERROR org.springframework.cloud.launcher.cli.LauncherCommand - Error running spring cloud
java.lang.reflect.InvocationTargetException: null
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.cloud.launcher.cli.LauncherCommand$LauncherOptionHandler.run(LauncherCommand.java:115)
	at org.springframework.boot.cli.command.options.OptionHandler.run(OptionHandler.java:84)
	at org.springframework.boot.cli.command.OptionParsingCommand.run(OptionParsingCommand.java:54)
	at org.springframework.boot.cli.command.CommandRunner.run(CommandRunner.java:219)
	at org.springframework.boot.cli.command.CommandRunner.runAndHandleErrors(CommandRunner.java:171)
	at org.springframework.boot.cli.SpringCli.main(SpringCli.java:63)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
15:13:42.198 [main] ERROR org.springframework.cloud.launcher.cli.LauncherCommand - Error running spring cloud
java.lang.reflect.InvocationTargetException: null
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.cloud.launcher.cli.LauncherCommand$LauncherOptionHandler.run(LauncherCommand.java:115)
	at org.springframework.boot.cli.command.options.OptionHandler.run(OptionHandler.java:84)
	at org.springframework.boot.cli.command.OptionParsingCommand.run(OptionParsingCommand.java:54)
	at org.springframework.boot.cli.command.CommandRunner.run(CommandRunner.java:219)
	at org.springframework.boot.cli.command.CommandRunner.runAndHandleErrors(CommandRunner.java:171)
	at org.springframework.boot.cli.SpringCli.main(SpringCli.java:63)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
Caused by: java.lang.IllegalStateException: Cannot download library for launcher org.springframework.boot.experimental:spring-boot-thin-launcher:jar:exec:1.0.1.RELEASE
	at org.springframework.boot.loader.wrapper.ThinJarWrapper$Library.download(ThinJarWrapper.java:194)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.getUrls(ThinJarWrapper.java:113)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.getClassLoader(ThinJarWrapper.java:105)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.launch(ThinJarWrapper.java:90)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.main(ThinJarWrapper.java:77)
	... 18 common frames omitted
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
Caused by: java.lang.IllegalStateException: Cannot download library for launcher org.springframework.boot.experimental:spring-boot-thin-launcher:jar:exec:1.0.1.RELEASE
	at org.springframework.boot.loader.wrapper.ThinJarWrapper$Library.download(ThinJarWrapper.java:194)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.getUrls(ThinJarWrapper.java:113)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.getClassLoader(ThinJarWrapper.java:105)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.launch(ThinJarWrapper.java:90)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.main(ThinJarWrapper.java:77)
	... 18 common frames omitted
```

**Comment by BoykoAlex** [comment]
Please try updating from Spring-IDE nightly: http://dist.springframework.org/snapshot/IDE/nightly/
I've added some logging to help me diagnose the problem. Wonder if you could try reproducing the problem again and send me the logs... Alternatively, you could try refreshing the `Local` target and see if you get any warning/errors in the error log view and just paste them here.

**Comment by Flaw101** [comment]
Updated both `CFT Integration`, and `AERI Integration` to versions `3.9.1.201708282033-CI-B1801` from `3.9.1.201708251858-CI-B1797` and no errors are thrown anywhere. This was using the packaged, `1.5.4.RELEASE` cli and a local `1.5.6.RELEASE` with the `1.3.4` Cloud CLI.

**Comment by Flaw101** [comment]
Sorry, just seen a couple - these may be completely unrelated given the unrelated stacks. 

```
null
Activator
Warning
Tue Aug 29 11:26:37 BST 2017
List services command output has more than one line:
 11:25:21.214 [main] DEBUG org.springframework.cloud.launcher.cli.LauncherCommand - resolved URIs [file:/Users/n0225684/.m2/repository/org/springframework/cloud/launcher/spring-cloud-launcher-deployer/1.3.4.RELEASE/spring-cloud-launcher-deployer-1.3.4.RELEASE.jar]
11:26:37.011 [main] ERROR org.springframework.cloud.launcher.cli.LauncherCommand - Error running spring cloud
java.lang.reflect.InvocationTargetException: null
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.cloud.launcher.cli.LauncherCommand$LauncherOptionHandler.run(LauncherCommand.java:115)
	at org.springframework.boot.cli.command.options.OptionHandler.run(OptionHandler.java:84)
	at org.springframework.boot.cli.command.OptionParsingCommand.run(OptionParsingCommand.java:54)
	at org.springframework.boot.cli.command.CommandRunner.run(CommandRunner.java:219)
	at org.springframework.boot.cli.command.CommandRunner.runAndHandleErrors(CommandRunner.java:171)
	at org.springframework.boot.cli.SpringCli.main(SpringCli.java:63)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
Caused by: java.lang.IllegalStateException: Cannot download library for launcher org.springframework.boot.experimental:spring-boot-thin-launcher:jar:exec:1.0.1.RELEASE
	at org.springframework.boot.loader.wrapper.ThinJarWrapper$Library.download(ThinJarWrapper.java:194)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.getUrls(ThinJarWrapper.java:113)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.getClassLoader(ThinJarWrapper.java:105)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.launch(ThinJarWrapper.java:90)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.main(ThinJarWrapper.java:77)
	... 18 common frames omitted
```
and additional,

```
null
Activator
Error
Tue Aug 29 11:26:37 BST 2017
NullPointerException: null

java.lang.NullPointerException
	at java.lang.String.contains(String.java:2133)
	at org.springframework.ide.eclipse.boot.core.cli.install.CloudCliInstall.getCloudServices(CloudCliInstall.java:90)
	at org.springframework.ide.eclipse.boot.dash.model.LocalBootDashModel.fetchLocalServices(LocalBootDashModel.java:293)
	at org.springframework.ide.eclipse.boot.dash.model.LocalBootDashModel.access$3(LocalBootDashModel.java:284)
	at org.springframework.ide.eclipse.boot.dash.model.LocalBootDashModel$6.run(LocalBootDashModel.java:313)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:56)

```

**Comment by BoykoAlex** [comment]
I have added a bit more logging. Lets try this again please. Update from the nightly build and check the log when `Local` is refreshed in Boot Dashboard. Thanks in advance!

**Comment by Flaw101** [comment]
Latest,

```
java.lang.NullPointerException
	at java.lang.String.contains(String.java:2133)
	at org.springframework.ide.eclipse.boot.core.cli.install.CloudCliInstall.getCloudServices(CloudCliInstall.java:94)
	at org.springframework.ide.eclipse.boot.dash.model.LocalBootDashModel.fetchLocalServices(LocalBootDashModel.java:293)
	at org.springframework.ide.eclipse.boot.dash.model.LocalBootDashModel.access$3(LocalBootDashModel.java:284)
	at org.springframework.ide.eclipse.boot.dash.model.LocalBootDashModel$6.run(LocalBootDashModel.java:313)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:56)
```
and 

```
List services command output has more than one line:
 10:47:49.647 [main] DEBUG org.springframework.cloud.launcher.cli.LauncherCommand - resolved URIs [file:/Users/n0225684/.m2/repository/org/springframework/cloud/launcher/spring-cloud-launcher-deployer/1.3.4.RELEASE/spring-cloud-launcher-deployer-1.3.4.RELEASE.jar]
10:49:13.967 [main] ERROR org.springframework.cloud.launcher.cli.LauncherCommand - Error running spring cloud
java.lang.reflect.InvocationTargetException: null
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.cloud.launcher.cli.LauncherCommand$LauncherOptionHandler.run(LauncherCommand.java:115)
	at org.springframework.boot.cli.command.options.OptionHandler.run(OptionHandler.java:84)
	at org.springframework.boot.cli.command.OptionParsingCommand.run(OptionParsingCommand.java:54)
	at org.springframework.boot.cli.command.CommandRunner.run(CommandRunner.java:219)
	at org.springframework.boot.cli.command.CommandRunner.runAndHandleErrors(CommandRunner.java:171)
	at org.springframework.boot.cli.SpringCli.main(SpringCli.java:63)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
Caused by: java.lang.IllegalStateException: Cannot download library for launcher org.springframework.boot.experimental:spring-boot-thin-launcher:jar:exec:1.0.1.RELEASE
	at org.springframework.boot.loader.wrapper.ThinJarWrapper$Library.download(ThinJarWrapper.java:194)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.getUrls(ThinJarWrapper.java:113)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.getClassLoader(ThinJarWrapper.java:105)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.launch(ThinJarWrapper.java:90)
	at org.springframework.boot.loader.wrapper.ThinJarWrapper.main(ThinJarWrapper.java:77)
	... 18 common frames omitted
```

**Comment by Jeff-Walker** [comment]
I have the problem also, only after upgrading to 3.9.0.

I found though, that if I create a new workspace, not importing any of my settings, that it works without a problem. That isn't a great solution for me since it took me months to get things how I like. :)

My stacktrace is the same, but for what it's worth, here are my session details.
`eclipse.buildId=3.9.0.201707061823-RELEASE-e47
java.version=1.8.0_51
java.vendor=Oracle Corporation
BootLoader constants: OS=win32, ARCH=x86, WS=win32, NL=en_US
Framework arguments:  -product org.springsource.sts.ide
Command-line arguments:  -os win32 -ws win32 -arch x86 -product org.springsource.sts.ide
`

**Comment by BoykoAlex** [comment]
Yes, this is because of old Boot CLI + Cloud CLI install pointed by the IDE. If you start working with a brand new workspace STS downloads 1.5.4 Boot CLI and 1.3.3 Cloud CLI automatically and that starts working.
I've pushed the fix for this issue i think. Should be available via the nightly builds update site: http://dist.springframework.org/snapshot/IDE/nightly/ Please check it out and let me know if it works for you. Thanks very much for testing this for me - it helps a lot!-)))

**Comment by Jeff-Walker** [comment]
@BoykoAlex That link downloads an empty file. Am I missing something?

**Comment by martinlippert** [comment]
The link above isn't a link to a download, but to a p2 repository. You could use this URL in the "Install New Software" dialog and install the updated features from there.

**Comment by Jeff-Walker** [comment]
Ok, that makes sense. I installed the Core/Spring IDE and the Spring boot integration. The error message is now gone.

One thing though, devtools still isn't reloading my app when classes change. Maybe this isn't related, but it seemed to stop when this error showed up, so I thought it was the cause of my devtools problem.

**Comment by martinlippert** [comment]
@Jeff-Walker I assume the devtools problem is a different one. Can you please open another issue for that and maybe point us to a sample project that helps us to reproduce the issue (and a few steps what you exactly did)? That would be great. Many many thanks upfront.
